architecture/2025-01-serverless-feature-spec.md

```diff
@@ -0,0 +1,279 @@
+- [Azure Container Instances (ACI)](https://learn.microsoft.com/en-us/azure/container-instances/): A serverless container runtime service that enables you to run containers on-demand without having to manage the underlying infrastructure.
+- [AWS Fargate](https://aws.amazon.com/fargate/): A serverless compute engine for containers that allows you to run containers without having to manage the underlying infrastructure.
+- [Azure Container Apps (ACA)](https://learn.microsoft.com/en-us/azure/container-apps/): A fully managed serverless container platform that abstracts away the need to manage containers altogether.
+- ACI [Container Groups](https://learn.microsoft.com/en-us/azure/container-instances/container-instances-container-groups): A container group is a collection of containers that get scheduled on the same host machine. The containers in a container group share a lifecycle, resources, local network, and storage volumes. It's similar in concept to a pod in Kubernetes.
+- ACI [nGroups](https://learn.microsoft.com/en-us/azure/container-instances/container-instance-ngroups/container-instances-about-ngroups): ACI feature that provides you with advanced capabilities for managing multiple related container groups, e.g. maintaining multiple instances, rolling upgrades, high availability, managed identity support, confidential container support, load balancing, zone rebalancing.
+- ACI [Confidential Containers](https://learn.microsoft.com/en-us/azure/container-instances/container-instances-confidential-overview): ACI feature that provides a secure enclave for your containerized applications to run in a confidential computing environment.
+- ACI [Spot Instances](https://learn.microsoft.com/en-us/azure/container-instances/container-instances-spot-containers-overview): ACI feature that allows you to run interruptible workloads at a reduced cost compared to the standard price by taking advantage of unused capacity in Azure datacenters.
+
+### Top level goals
+<!-- At the most basic level, what are we trying to accomplish? -->
+1. Genericize the Radius platform to support deployment of applications on serverless (and other) compute platforms beyond Kubernetes.
+1. Enable developers to deploy and manage serverless containers in a way that is consistent with the existing Radius environment and application model.
+1. Make all Radius features (Recipes, Connections, App Graph, UDT, etc.) available to engineers building applications on Radius-enabled supported serverless platforms.
+1. Radius support for unopinionated or process-driven serverless container runtimes (e.g. AWS Elastic Container Service, Azure Container Instances, Azure Container Apps, AWS Fargate).
+
+### Non-goals (out of scope)
+<!-- What are we explicitly not trying to accomplish? -->
+1. Hosting the Radius control plane on serverless (and other) compute platforms outside of Kubernetes.
```

Comment:
This is really interesting - users will be required to have Kubernetes even when they do not want to deploy to Kubernetes. Is the reason for this to reduce the scope of the effort? Or is there another reason for keeping the control plane on Kubernetes only?

Reply:
> Is the reason for this to reduce the scope of the effort?
Yes, this is precisely the reason. Hosting Radius itself on serverless or other compute platforms will be a future feature that follows this.